### PR TITLE
Stricten typescript in protocol-authorization.ts

### PR DESCRIPTION
**Pull request by diehuxx**
Taking a small step towards getting this repo onto strict typescript rules (#261). Starting with the file `protocol-authorization.ts` in preparation for #242.

Typescript doesn't support per-file strictness out of the box, but `typescript-strict-plugin` does. In future PRs, if you edit a file, please add it to the list of strict files in the tsconfig 🙏.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -115,6 +115,7 @@
     "search-index": "3.4.0",
     "sinon": "13.0.1",
     "typescript": "4.8.2",
+    "typescript-strict-plugin": "^2.1.0",
```
Comment (thehenrytsai) [inline package.json:118]: Are you adding this dependency temporarily until we have moved everything to stricter checks?

An alternative approach to make progress for #261 would be to turn `"strict": true` on in `tsconfig` and fix what you can/want then comment out the rule again.

I just do now want:
1. for this dependency to be permanent; and
1. for us to add every file in the `paths` in both `tsconfig` files in order to complete #261, turning on the strict rule itself seems to be easier to be sure and know the actual progress.

File: src/core/protocol-authorization.ts [hunk 2]
```diff
@@ -147,7 +155,11 @@ export class ProtocolAuthorization {
       }
 
       // else we keep going down the message chain
-      allowedRecordsAtCurrentLevel = allowedRecordsAtCurrentLevel[currentRecordType].records;
+      const nextAllowedRecords = allowedRecordsAtCurrentLevel[currentRecordType].records;
+      if (nextAllowedRecords === undefined) {
+        throw new Error(`No record types found in schema '${currentRecordSchema}' in structure level ${currentMessageIndex}`);
```
Comment (thehenrytsai) [inline src/core/protocol-authorization.ts:160]: Seems legitimate case caught by TS. Please add a test for it.

File: package-lock.json [hunk 3]
```diff
@@ -1872,6 +1873,27 @@
         "node": ">=12"
       }
     },
+    "node_modules/cli-cursor": {
+      "version": "3.1.0",
+      "resolved": "https://artifactory.global.square/artifactory/api/npm/square-npm/cli-cursor/-/cli-cursor-3.1.0.tgz",
```
Comment (thehenrytsai) [inline package-lock.json:1878]: Sanity: are you sure all these tens of `https://artifactory.global.square` square specific reference won't break install? They break for me.

File: src/core/protocol-authorization.ts [hunk 4]
```diff
@@ -207,7 +219,10 @@ export class ProtocolAuthorization {
       allowedActions = allowRule.anyone.to;
     } else if (allowRule.recipient !== undefined) {
       allowedActions = allowRule.recipient.to;
-    } // not possible to have `else` because of same check already done by verifyAllowedRequester()
+    } else {
+      // not possible to have `else` because of same check already done by verifyAllowedRequester()
+      allowedActions = [];
+    }
```
Comment (thehenrytsai) [inline src/core/protocol-authorization.ts:225]: instead just do:
`let allowedActions: string[] = [];`

File: src/core/types.ts [hunk 5]
```diff
@@ -65,5 +70,5 @@ export type LT = ({ lt: string } & { lte?: never }) | ({ lt?: never } & { lte: s
 export type RangeFilter = (GT | LT) & Partial<GT> & Partial<LT>;
 
 export type Filter = {
-  [property: string]: EqualFilter | OneOfFilter | RangeFilter
+  [property: string]: DwnInterfaceFilter | DwnMethodFilter | EqualFilter | OneOfFilter | RangeFilter
```
Comment (thehenrytsai) [inline src/core/types.ts:73]: Sanity: is this really necessary? IMO this is much more confusing than what it was before.

File: src/core/protocol-authorization.ts
Comment (thehenrytsai) [inline src/core/protocol-authorization.ts:1]: The code coverage dropped from 100% to ~92%. In general, when this happens it really begs a closer look, typically it means either the code added is redundant (unreachable), or there are missing tests. Either way makes the case to bring the code coverage back to 100%.

File: package.json (hunk 1 above)
Comment (diehuxx) [reply package.json:118]: Yes, the intention was to use this temporarily until we finish migrating to strict. I don't feel very strongly about this approach, so turning on `"strict": true` while developing works for me.

File: package-lock.json (hunk 3 above)
Comment (diehuxx) [reply package-lock.json]: Ah, sorry didn't realize it was adding artifactory stuff. In any case, I removed the package from this PR

File: src/core/types.ts (hunk 5 above)
Comment (diehuxx) [reply src/core/types.ts]: Removed

File: src/core/protocol-authorization.ts
Comment (diehuxx) [reply src/core/protocol-authorization.ts:1]: Took a closer look, and you're absolutely right. I added a test and made a slight change to how we're handling an error. Let me know what you think.

**Pull request merged**